dali/kernels/imgproc/convolution/laplacian_cpu.h

```diff
@@ -0,0 +1,311 @@
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+
+namespace laplacian {
+
+using namespace conv_transform;  // NOLINT
+
+/**
+ * @brief Computes convolution to obtain partial derivative in one of the dimensions.
+ * Convolution consits of `axes` windows, each to convolve along one dimension of the input data,
+ * where `deriv_axis`-th window is supposed to compute partial derivative along that axis,
+ * whereas the remaining windows should perform smoothing. It is assumed that smoothing
+ * window of size 1 must be equal to `[1]`, this way, if window sizes in non-derivative directions
+ * are one, the smoothing convolutions can be skipped and only a single one-dimensional
+ * convolution in derivative direction is performed.
+ */
+template <typename Out, typename In, typename W, int axes, int deriv_axis,
+          bool has_channels = false, typename T = conv_transform::TransScaleSat<Out, W>>
+struct PartialDeriv {
+  using MultiDimConv = SeparableConvolutionCpu<Out, In, W, axes, has_channels, T>;
+  static constexpr int ndim = MultiDimConv::ndim;
+  using SingleDimConv = ConvolutionCpu<Out, In, W, ndim, deriv_axis, has_channels, T>;
+
+  bool IsNoSmooting(const std::array<int, axes>& window_sizes) {
+    for (int axis = 0; axis < axes; axis++) {
+      if (axis != deriv_axis && window_sizes[axis] != 1) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    no_smoothing_ = IsNoSmooting(window_sizes);
+    if (no_smoothing_) {
+      if (!single_dim_impl_) {
+        multi_dim_impl_.reset();
+        single_dim_impl_ = std::make_unique<SingleDimConv>();
+      }
+      return single_dim_impl_->Setup(ctx, in_shape, window_sizes[deriv_axis]);
+    }
+    if (!multi_dim_impl_) {
+      single_dim_impl_.reset();
+      multi_dim_impl_ = std::make_unique<MultiDimConv>();
+    }
+    return multi_dim_impl_->Setup(ctx, in_shape, window_sizes);
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> &out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const T& transform = {}) {
+    if (no_smoothing_) {
+      single_dim_impl_->Run(ctx, out, in, windows[deriv_axis], transform);
+    } else {
+      multi_dim_impl_->Run(ctx, out, in, windows, transform);
+    }
+  }
+
+  bool no_smoothing_;
+  std::unique_ptr<MultiDimConv> multi_dim_impl_;
+  std::unique_ptr<SingleDimConv> single_dim_impl_;
+};
+
+
+/**
+ * @brief Provides Laplacian specializations for 2 and 3 dimensional data.
+ *  Run methods expect additional ``acc`` buffer that will be used to accumulate
+ *  partial derivatives. If ``Intermediate`` and ``Out`` are the same type,
+ *  the ``acc`` and ``out`` can be the same tensor.
```

Comment:
What would happen, if `Intermediate != Out` and user provides the same tensor to `acc` and `out`?

Reply:
Well, without some explicit casts it is not possible, as the tensors ``acc`` and ``out``  have different types.